### ISSUE TITLE
Improve telemetry package to better support additional sink types

### ISSUE THREAD

**Issue by martincapello** [spiffe/spire#547]
Right now the agent uses a telemetry package which implements a Sink interface whose current implementation keeps the data in memory and dumps it on the agent's log when it receives a SIGUSR1 signal. This doesn't seem to be flexible enough, so an improvement on SPIRE's code base is required in order to support (via configuration option) a StatsD backend, and set the stage to be able to easily add support for new backends in the future.

**Comment by eicnix** [comment]
I would like to see a Prometheus metric endpoint that exposes these metrics on the server and agent. A lot of Kubernetes deployments use Prometheus as their primary monitoring tool and Prometheus support would improve the Kubernetes deployment of SPIRE.

**Comment by martincapello** [comment]
Currently, under the hood, SPIRE's telemetry package is using https://github.com/armon/go-metrics, which in turn supports a PrometheusSink. So I believe once this work is done, shouldn't be too difficult to add Prometheus support.

**Comment by martincapello** [comment]
I propose having a new configuration section on the server and agent config files as follows:
```
    telemetry {
        statsd {
            addr = "somehost:8000"
        }
        statsite {
            addr = "otherhost:9000"
        }
        prometheus {
            expiration = "60"
        }
    }
```
This section would be inside of agent {} or server {} sections of their corresponding config files.
Each subsection inside telemetry{} its named as the sink implementation we want to use, and the properties inside define their specific configuration. If more than one sink is specified the metrics will  fanout to all of them. 
If no telemetry{} section or sinks are specified, the behavior must be the same as it is currently.
Then in github.com/spiffe/spire/pkg/common/telemetry/telemetry.go the SinkConfig struct has to be updated to accept the configuration of each supported sink, and the NewSink must be changed to instantiate each sink and add them to the FanoutSink as needed.

**Comment by evan2645** [comment]
SGTM

might be worth getting config format review from @ajessup I know he has been mulling config improvements

**Comment by ajessup** [comment]
I think as far as OSS work is concerned this proposal is fine (also, yay - prometheus!). My suspicion here is that most folks will use only 1-2 telemetry systems (though @eicnix feel free to disagree!) so any redundant config is less of a concern.

**Comment by bradleyjames** [comment]
> My suspicion here is that most folks will use only 1-2 telemetry systems

What is a system in this statement?  We have an internal API we publish metrics to.  Will this design allow that?

**Comment by bradleyjames** [comment]
If we write a custom sink will we be able to configure it via the above mechanism?

**Comment by martincapello** [comment]
@bradleyjames no, as you probably noticed already, you won't be able because each new sink needs new code to be added to support it. We are happy to add support for popular sink types when we can, if you have an in-house metric system perhaps you can use logstash, collectd, or a custom shim to massage statsd data into whatever system you have.

**Comment by bradleyjames** [comment]
Is a plugin API off the table?

**Comment by martincapello** [comment]
Most people with in-house metrics systems have ways to ingest metrics from popular collection protocols like statsd or prometheus scraping, as getting custom metrics integrated into every open source project is very difficult.
We could extend our plugin system to support telemetry plugins, however significant changes would be required and the current system is good enough for the majority of the users, so we are not planning to undertake that work any time soon.

**Comment by evan2645** [comment]
Fixed in #817 

If you have a use case that this doesn't address, please open a new issue
